### ISSUE TITLE
Unable to use navigation hotkeys in Notes window

### ISSUE THREAD

**Issue by ptorpey** [jcsteh/osara#825]
I started using the Notes window available in Reaper.  The action to open the Notes window I am using is:
SWS/S&M: Open/close Notes window

From the actions list it appears that there might also be lots of other notes windows available like notes for items, notes for tracks, etc.

The problem is that when in the Notes dialog, the standard navigation hotkeys for navigating through a document are being captured as if the TCP was open.

For example, pageup/pagedown move back/forward a measure instead of going back/forward through the document.  Similar for the Hoem and End keys, control+home and control+end, etc.

Would be niced if such navigation commands worked in such editable dialogs or fields.

--Pete

**Comment by ScottChesworth** [comment]
It'd benefit more users if this was shored up by Cockos. I'll pass it along.

**Comment by ScottChesworth** [comment]
@ptorpey, can you re-check page keys and tell me whether they behave differently with and without Snowman's scripts running? With NVDA here I'm not getting timeline navigation as you've described, but I've found another non-standard behaviour where I can't scroll between pages with arrow keys. That's an interesting one because I'm instantly seeing uses for it, but decades of convention says no.

**Comment by ptorpey** [comment]
Okay, I should have thoroughly tested all of the navigation keys instead of just testing a few and assuming they were all broken.

 

*	pageUp/PageDown seem to work properly for moving up and down the Notes content with and without Jim’s scripts running as would be expected in an edit field.  There is no feedback from JAWS, however, that we have moved up/down as there usually is.

*	Home and End properly go to the beginning of line and end of line, respectively in the Notes as would be expected in an edit field.  No feedback from JAWS about actually having moved, however, as there usually is, with or without Jim’s scripts.
*	Up/Down/Left/Right arrow keys seem to navigate properly in the edit field of Notes as expected.
*	Control+Home and control+End do not move to the top and bottom of the Notes dialog as one would like.  Rather they are still activating the move to beginning and end of project, respectively with the corresponding feedback from Osara.  Again, same behavior with and without Jim’s JAWS scripts.  So this is broken.
*	Control+left/right arrows do not move to move a word at a time as expected.  So this is broken also.  This is still tied to moving to previous/next item.
*	Selection with shift+left/right arrow works to select a character at a time, but control+shift+left/right arrow does not and is still tied to selecting contiguous items.
*	Selection with shift+PageUp/PageDown seems to work but is not reported using Jim’s script.  Selection and selected text is reported using JAWS default.

 

In summary  it seems that navigation/selection commands in the Notes edit field seem to work except when using special keys such as control.

 

Don’t know if I missed any navigation commands.

 

There also seem to be notes for other things such as object.  For example, the action:

Item: Show notes for items...

Presumably shows notes for an item.  Interestingly, the control key + nav navigation hotkey seem to work okay in at least this notes dialog.  I haven’t tested or searched for all of the other types of Notes dialogs.

 

As an aside, the main Notes dialog could be useful for people when they want to keep notes about a particular project linked to the project and readily accessible.  I set the hotkey alt+Y to the action:

SWS/S&M: Open/close Notes window

To get to this quickly.  I was surprised that none of the types of Notes views were available in the “View” pull-down menu.  Maybe a suggestion to Cockos.

 

--Pete

 

 

 

 

--Pete

*	 
*

**Comment by ScottChesworth** [comment]
I think what's going on is, mappings that aren't global scope aren't making it through to those edit fields as expected. When I unassigned those actions, navigation improved. Obviously that's not a viable solution. Conversation with Cockos is already in progress. Will leave this open and report back when I know more.

**Comment by jcsteh** [comment]
Do those keys work in REAPER multi-line edit fields like the one in the FX dialog? It's slightly possible that this is deliberate (or at least a side effect of intended behaviour) in SWS.

**Comment by ScottChesworth** [comment]
They don't work there unless unassigned nope. Also, neither does Enter work there here on Windows even though it's identified as a multi-line field.
Also also, you're on holiday, stop being useful! :)

**Comment by ScottChesworth** [comment]
So after consulting/confusing Justin and a rogue appearance of an off-duty Jamie, it turns out this was us all along. DOH! Seems someone got a bit carried away with the scope setting on our key map. I'm gonna go through and reassign a bunch of stuff with Normal scope which should fix the majority of navigation issues raised here. Cockos are looking at Control+Backspace to erase by word because that doesn't work yet. Gonna need to get some testing on these scope adjustments to make sure we're not impacting anything by enabling better navigation.

**Comment by ptorpey** [comment]
Thanks for following up with Justin.  Do you think it is worth adding an Osara defined hotkey to open up some form of the Notes?  Is that a feature that many people would make use of (or do they even know about it)?

 

I was amazed to learn just how many variants of notes windows there are.  In addition to the one that raised this issue:

SWS/S&M: Open/close Notes window

there are also the following:

: SWS/S&M: Open/close Notes window (extra project notes)
SWS/S&M: Open/close Notes window (extra project notes), State: off
: SWS/S&M: Open/close Notes window (global notes)
: SWS/S&M: Open/close Notes window (item notes)
: SWS/S&M: Open/close Notes window (marker names)
: SWS/S&M: Open/close Notes window (marker subtitles)
: SWS/S&M: Open/close Notes window (marker/region names)
: SWS/S&M: Open/close Notes window (marker/region subtitles)
: SWS/S&M: Open/close Notes window (project notes)
: SWS/S&M: Open/close Notes window (region names)
: SWS/S&M: Open/close Notes window (region subtitles)
: SWS/S&M: Open/close Notes window (track notes)

 

I don’t know if any of these are tied together or if they are completely separate.  I also have no idea how a sighted person might even know of the existence of some of these specialized notes.  Not many of these appear in any of the Reaper menus.

 

--Pete

 

 

--Pete

**Comment by ScottChesworth** [comment]
Way too many variants and not enough data on adoption for us to map IMO.

**Comment by ptorpey** [comment]
Way too many variants and not enough data on adoption for us to map IMO.

[PT] Agreed.  I do wonder though, whether many people are aware of the notes option at all.  Personally, I am happy to use the one Notes option since it is one easy place to go to find notes about a project.  I did notice that there is an option in the Project Settings dialog to have the Notes dialog open automatically when a project loads if a user is interested in that.  People can also map an unused keystroke (as I do) to toggle that notes dialog open/closed.  If you get the keystroke navigation stuff straightened out I think we can close this issue.  Thanks again.

 

--Pete

**Comment by ptorpey** [comment]
If there winds up being a Reaper dev build with some of these fixes to be able to use navigation keys in the notes dialog, let me know and I’ll beat on it.  I’m making more and more use of that notes dialog.  Very handy.

 

Thanks.

 

--Pete

**Comment by ScottChesworth** [comment]
we're not waiting on a dev build from Cockos, it's changes to our Windows key map that are required to shore this one up. Someone's gotta go through and reset all bindings relevant to text navigation, making sure that they're standard scope. I'll try to get to it later today.

**Comment by ptorpey** [comment]
My misunderstanding.  I’ll look for a notice on GITHub.  Thanks.

 

--Pete

**Comment by RDMurray** [comment]
@ScottChesworth I got the latest snapshot and can't find any global keys in the keymap. They are listed with "(global)" in the page generated by shift+F1.

I think this is an SWS bug. If you access project notes from the project properties dialog, or item notes using the built-in reaper action all the keys listed above work as expected.

The rest of the notes actions bring up a window provided by SWS which have the problem.

**Comment by ScottChesworth** [comment]
Yeah, realised I was on the wrong track tinkering with this earlier, just haven't had time to update the thread. Here, it seems to be any combination involving Control that doesn't behave as expected. Could that have anything to do with the recent change around how the control key is handled? I didn't really grasp what that PR changed.

**Comment by ptorpey** [comment]
If you access project notes from the project properties dialog, or item notes using the built-in reaper action all the keys listed above work as expected.

The rest of the notes actions bring up a window provided by SWS which have the problem.

[PT] Verified here.  Control+Keys do not work as they should when bringing up notes using the SWS action, but editing notes in the Project Settings / Notes dialog does work fine.

 

Two additional comments:

1.	When navigating the Project Settings, it would be nice if control+tab and control+shift+tab would move to the next and previous tabs as with most other multi-tab dialogs.  Currently it takes lots of tabbing to move from tab to tab dialog.  Is this something Cockos could fix?
2.	I noticed that although control+keys work in the Project Settings / Notes dialog as they should without the JAWS scripts running, the behavior is not working correctly with the JAWS scripts running.  Once we figure out the SWS problem with notes and if it can be rectified or not, I can get in touch with Jim about this.

 

--Pete

**Comment by RDMurray** [comment]
Since this is an SWS bug, should it be fixed in SWS? This would mean waiting for an SWS release. Alternatively we could work around it in Osara, in the same way as the transient detection settings dialog.

Either way, we need a list of keys that should be handled by the dialog. Let me know if I missed any:
* Control+left/right/home/end
* Control+Shift+left/right/home/end

Control+backspace doesn't work when handled by the dialog for some reason. This is also the case for the notes in project properties.

**Comment by ptorpey** [comment]
Additional keys that should work in an edit window.  I believe these work when not using the JAWS scripts but do not work if using the JAWS scripts:

*	PageUp/PageDown
*	Home/End (without the modifiers to get to the beginning and end of a line)
*	Shift+PageUp/PageDown
*	Shift+Home/End
*	And, of course, hotkeys such as control+C/V/X

 

As I said, once we think we have a fix or direction , I can talk to Jim about those.

 

I didn’t realize SWS was ever updated any more.  The last update I see on the Reaper dev page is:

2019-08-09 

Is SWS a dead duck in the water?  Who actually maintains that these days?

**Comment by ptorpey** [comment]
I wonder since Cockos does support project notes if it is worth asking them to support:

1.	Opening the project notes via an action (rather than having to dig down into the Project Settings / Notes dialog), and
2.	Offering the user’s choice of editor for opening notes (as they do with editing Actions, etc.)

 

The various other SWS open notes actions are for things like item notes, marker and region notes, etc., and are likely of less utility.  But the project notes can be very useful for keeping notes about a project with the project.

 

--Pete

**Comment by ScottChesworth** [comment]
Support for Control+Backspace needs to come from Cockos. Already raised it with Justin.

**Comment by RDMurray** [comment]
Ive reported this issue to SWS and will probably submit a fix as it is quite straight forward. BTW it looks like SWS is preparing for a release soon. It would be good to get this fix in before the release, but I doubt that is going to happen given that they haven't merged anything since may.

**Comment by RDMurray** [comment]
Control+x/c/v already work. I added Control+Z to the list above.

**Comment by ScottChesworth** [comment]
Maybe tag cfillion and nofishonfriday when you submit the PR. Both good dudes who know what OSARA/accessibility is about.

**Comment by RDMurray** [comment]
@cfillion is already looking into it. He has it working on macOS as well apart from command+Z. 
reaper-oss/sws#1721

**Comment by ptorpey** [comment]
Just tested here on Windows 10.  Looks good with JAWS default config.  Works pretty well with Jim’s scripts, but not exactly like default behavior.  I’ll contact Jim to see if he can fix that.

 

Good job.  Thanks for following up and contacting the SWS team.  I like using this notes feature.  Great to toggle the notes window on/off with a hotkey for quick reference inside a project.

 

--Pete

**Comment by Lo-lo78** [comment]
Very interesting this thread.
I was thinking of a problem. The OSARA keymap is very large and the free shortcuts are less and less.
Is it possible that when we select a track or an item or a marker the shortcut adjusts to the focus?
A bit like the delete key.
Otherwise we will occupy 12 shortcuts to have all the notes at your fingertips.
Maybe we need a new issue for this. ;))

**Comment by ScottChesworth** [comment]
Nice idea. Best requested from Cockos IMO, it would be useful core functionality. I'll ask them.

**Comment by RDMurray** [comment]
I don't think Cockos can help here; every action apart from "Item: Show notes for items..." is provided by SWS.

**Comment by RDMurray** [comment]
This is fixed in the [latest dev build of SWS](https://www.sws-extension.org/download/pre-release/)


Do we still need a workaround in Osara, or can this be closed?

**Comment by ptorpey** [comment]
I cannot test this on  Mac, but on Windows it seems to be working fine using the default JAWS configuration.  I’ve asked Jim to look at patching up the JAWS scripts so that navigation in the Notes dialog also works with his scripts running.

 

 

So I think this can be closed unless we hear from Mac users about an issue.  I’ll keep you posted if Jim does address this with the JAWS scripts.

 

Thanks.  I find the Notes to be very handy.

 

--Pete

**Comment by ScottChesworth** [comment]
Jen checked it over on Mac a while back, all seemed ok.

**Comment by RDMurray** [comment]
The thing is it still isn't in a stable release of SWS and I have no idea when that will happen. The latest release was 2 years ago now.

**Comment by ScottChesworth** [comment]
My thinking is that it's not a particularly widely used set of features, it won't break my back or result in a ton of extra traffic if we just point anyone who notices things aren't quite right to an SWS dev build. Of course if you want to fix it in OSARA too, knock yourself out.

**Comment by ptorpey** [comment]
Well I’ll trust that this version is pretty stable and save the link for others who might want to use the Notes feature.

 

I wonder why they don’t just merge these once in a while.  I noticed the version on the Reaper dev page is really old.

 

Oh well.  The pros and cons of open source!

 

--Pete

**Comment by ptorpey** [comment]
Do you think most people even know about the feature?  I’m guessing not.

 

--Pete

**Comment by ScottChesworth** [comment]
I know of a few folks using Chessel's scripts to report notes, only a few though.

**Comment by ptorpey** [comment]
Chessel’s scripts?  What are those and where can I take a look at them.

 

Thanks.

 

--Pete

**Comment by ScottChesworth** [comment]
He wrote scripts to speak track and item notes. They're in the Dropbox, go to Scripts, ReaScripts, Chessel.

**Comment by ptorpey** [comment]
Fixed in Jim’s latest scripts to be released soon.  The SWS beta must also be installed.  Would be good if they had a released version so that it showed up as the latest version on the Cockos development web site to avoid future confusion.

 

--Pete
